src/util/http.go

```diff
@@ -0,0 +1,220 @@
+package util
+
+import (
+	"bytes"
+	"encoding/json"
+	"encoding/xml"
+	"github.com/ImpactDevelopment/ImpactServer/src/util/mediatype"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"strconv"
+	"strings"
+)
+
+const userAgent = "ImpactServer"
+
+// HTTPRequest wraps http.Request so that we can provide custom methods
+type HTTPRequest struct {
+	Req *http.Request
+
+	// client is the http.Client which will do the request
+	//
+	// it is set here so it can be overridden by tests
+	client *http.Client
+}
+
+// HTTPResponse wraps http.Response so that we can provide custom methods
+type HTTPResponse struct {
+	Resp *http.Response
+	Body []byte
+}
+
+// NewRequest wraps http.NewRequest but returns HTTPRequest instead of http.Request. The ImpactServer User Agent is automatically added
+//
+// You probably want to use one of its wrappers like GetRequest or JSONRequest instead
+func NewRequest(method, url string, body io.Reader) (*HTTPRequest, error) {
+	request, err := http.NewRequest(method, url, body)
+	if err != nil {
+		return nil, err
+	}
+
+	request.Header.Set("User-Agent", userAgent)
+
+	return &HTTPRequest{
+		Req:    request,
+		client: http.DefaultClient,
+	}, nil
+}
+
+// GetRequest returns a HTTPRequest using method GET with no body
+func GetRequest(address string) (*HTTPRequest, error) {
+	return NewRequest(http.MethodGet, address, nil)
+}
+
+// JSONRequest returns a HTTPRequest using method POST with a JSON marshalled body
+func JSONRequest(address string, body interface{}) (*HTTPRequest, error) {
+	post, err := json.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+
+	request, err := NewRequest(http.MethodPost, address, bytes.NewReader(post))
+	if err != nil {
+		return nil, err
+	}
+
+	request.setContentType(mediatype.JSON)
+	request.setLength(len(post))
+
+	return request, nil
+}
+
+// XMLRequest returns a HTTPRequest using method POST with an XML marshalled body
+func XMLRequest(address string, body interface{}) (*HTTPRequest, error) {
+	// Strip the trailing newline, since we add our own later.
+	doctype := xml.Header[:len(xml.Header)-1]
+	return XMLRequestWithDoctype(address, doctype, body)
+}
+
+// XMLRequestWithDoctype returns a HTTPRequest using method POST with an XML marshalled body with the specified doctype
+func XMLRequestWithDoctype(address, doctype string, body interface{}) (*HTTPRequest, error) {
+	postBody, err := xml.Marshal(body)
+	if err != nil {
+		return nil, err
+	}
+	post := bytes.Join([][]byte{[]byte(doctype), postBody}, []byte("\n"))
```

Comment:
w--w--what

`append(append(doctype, postBody...), []byte("\n")...)` is more standard i think

Reply:
I'm not appending `\n` to the end, I'm joining doctype and postBody with a `\n` in between.

Reply:
whAT the fUCK that's completely unclear!